### ISSUE TITLE
[REVIEW]: Delve: Neural Network Feature Variance Analysis

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3992]
**Submitting author:** @justinshenk (<a href="http://orcid.org/0000-0002-0664-7337">Justin Shenk</a>)
**Repository:** <a href="https://github.com/delve-team/delve" target ="_blank">https://github.com/delve-team/delve</a>
**Version:** v0.1.49
**Editor:** @rkurchin
**Reviewer:** @sambitdash, @dionhaefner
**Archive:** <a href="https://doi.org/10.5281/zenodo.5865465" target="_blank">10.5281/zenodo.5865465</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/364a58b2dc11293da55bd26af2d3f503/status.svg)](https://joss.theoj.org/papers/364a58b2dc11293da55bd26af2d3f503)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/364a58b2dc11293da55bd26af2d3f503"><img src="https://joss.theoj.org/papers/364a58b2dc11293da55bd26af2d3f503/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/364a58b2dc11293da55bd26af2d3f503/status.svg)](https://joss.theoj.org/papers/364a58b2dc11293da55bd26af2d3f503)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/delve-team/delve" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@sambitdash & @dionhaefner, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @rkurchin know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @sambitdash
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/delve-team/delve">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/delve-team/delve">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@justinshenk) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @dionhaefner
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/delve-team/delve">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/delve-team/delve">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@justinshenk) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @sambitdash, @dionhaefner it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1107

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.10 s (528.7 files/s, 50917.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          19            491            579           2291
reStructuredText                21            296            251            361
TeX                              1             18              0            231
Markdown                         2             61              0            214
YAML                             7             12              6            184
DOS Batch                        1              8              1             26
make                             1              4              7              9
Jupyter Notebook                 1              0             52              2
-------------------------------------------------------------------------------
SUM:                            53            890            896           3318
-------------------------------------------------------------------------------


Statistical information for the repository 'c6b96ae439c5a7d5fac40dc4' was
gathered on 2021/12/14.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Cypher42                        64          3500           2894           46.33
Johan Byttner                    3           203            127            2.39
Justin Shenk                   161          4437           1855           45.59
MLRichter                       18           350            205            4.02
Wolf Byttner                     5           171             26            1.43
mmarcinkiewicz                   1            12             20            0.23

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Johan Byttner               155           76.4         18.9                0.65
Justin Shenk               1911           43.1          8.9               11.25
MLRichter                  1178          336.6         19.3                6.71
Wolf Byttner                117           68.4         13.1                1.71
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/wacv.2018.00097 is OK
- 10.1007/s11263-019-01228-7 is OK
- 10.5281/zenodo.5233860 is OK
- 10.1007/978-3-030-86340-1_11 is OK
- 10.1109/cvpr.2016.319 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by sambitdash** [comment]
Hi @rkurchin,

I reviewed the software paper. 

There are some typos. They are identified and a PR is submitted for easier recognition and fixing. https://github.com/delve-team/delve/pull/53.

Although the paper talks about the limitations of the other techniques, it does not have a reference to a numerical comparison of those techniques with the proposed technique. If that has not been carried out, maybe that should be mentioned as a scope of further work in the paper.

**Comment by justinshenk** [comment]
Hi @sambitdash,
Thank you for your valuable suggestions and improvements. The PR has been merged and a note added in https://github.com/delve-team/delve/commit/811756520cbfd8dce4427c53203ac193f61a94d1 regarding the scope of further work.

**Comment by justinshenk** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by dionhaefner** [comment]
Hey @justinshenk, congratulations on the nice library you and your collaborators have built!

I opened some issues related to the code and docs. On top of that, here are some points I had regarding the paper:

- I think you do a good job of summarizing the status quo, but what I find missing is a clear statement on what delve is trying to be. Is delve a tool to monitor layer saturation? Is it a general monitoring tool for layer statistics? From the REAMDE and docs I thought it was the former, but the paper makes it sounds like the goal is the latter.
- If your goal for delve is to become a general one-stop tool for layer statistics you will need to lay out your plans in a lot more detail. What are the next steps towards this? Adding more layer types can't be the only thing left to do. (I would probably suggest you frame this mostly as a tool for layer saturation though, and the rest as the cherry on top.)
- The section "Eigendecomposition of the feature covariance matrix" does nothing for me personally, because it is too brief to understand what's going on. Either remove it and refer to the paper or at least explain all quantities and operators in the equation.

**Comment by whedon** [comment]
:wave: @sambitdash, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @dionhaefner, please update us on how your review is going (this is an automated reminder).

**Comment by MLRichter** [comment]
Hello,
sorry for the delayed answer during Christmastime.
Hope you all had a great holiday season!
Thank you for your feedback, I have compiled some answers to your remarks:

1. What is delve trying to be?
It is primarily a tool for monitoring saturation and derived information like the intrinsic dimensionality and some visualizations, and this is its primary goal.
It CAN be used for experiment control as well. However, this can be considered as a freebie, that just comes with the logic necessary for the former. There are more advanced solutions for pure experiment control and monitoring like MLFlow and the stuff from weights and biases for example.
This is also the reason why we work with the writer-patterns, which makes it easy to hook delve into any data sink and by extension monitoring system, which we think is important for easy integration into existing code.
That being said, I agree that it should be clarified in the text, that we are mainly providing interesting statistics based on the spectrum of the latent representations.

2. Regarding the topic "Eigendecomposition of the feature covariance matrix"
I agree with the reviewer that this chapter is either undersized or unnecessary.
I think it is essential to elaborate on the way the covariance matrix is computed, since its memory and computationally efficient computation during training is one of the selling points of delve.
Thus, I would like to expand in this chapter and elaborate the computation of the covariance matrix in greater detail. Furthermore highlighting the benefits of our approximation method when it comes to practical usage.

**Comment by MLRichter** [comment]
Happy New Year everyone,
This pull request, contains the reworked text Eigendecomposition:
https://github.com/delve-team/delve/pull/60
Please let me know if this is sufficient!

**Comment by MLRichter** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by dionhaefner** [comment]
Thanks @MLRichter, the section is much better now!

The remaining to-do list from my side is this:

- [ ] Clarify in the paper what Delve is trying to be, i.e.
  
   > primarily a tool for monitoring saturation and derived information like the intrinsic dimensionality and some visualizations

- [ ] Address the issues on the Delve repository.
- [ ] I noticed quite a few typos in the new section, so please run the manuscript through a spell checker or ask a colleague to provide some copy-editing.

After that we're good to go!

**Comment by sambitdash** [comment]
Thanks a lot for all the updates. I shall be reviewing the next round over the weekend and providing my feedback.

**Comment by MLRichter** [comment]
Hi,
just giving you a brief update.
Currently, work is almost finished on these three PRs: 
https://github.com/delve-team/delve/pull/61 addressing Issue: https://github.com/delve-team/delve/issues/54
https://github.com/delve-team/delve/pull/63 addressing Issues: https://github.com/delve-team/delve/issues/57 and https://github.com/delve-team/delve/issues/58
https://github.com/delve-team/delve/pull/61 addressing the issues regarding the written paper. 

https://github.com/delve-team/delve/issues/59   is the last open issue, will be addressed next?

**Comment by MLRichter** [comment]
Update: All PRs have been merged.

**Comment by dionhaefner** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by justinshenk** [comment]
@whedon generate pdf

**Comment by justinshenk** [comment]
Merged, thank you for the helpful revisions!

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by rkurchin** [comment]
@sambitdash and @dionhaefner, it seems you've both completed your checklists. Could you confirm that you're now happy to accept this submission?

**Comment by dionhaefner** [comment]
There's still delve-team/delve#59, but if the authors prefer to address that later it's fine with me.

**Comment by sambitdash** [comment]
> There's still [delve-team/delve#59](https://github.com/delve-team/delve/issues/59), but if the authors prefer to address that later it's fine with me.

@dionhaefner, agree with you.

**Comment by rkurchin** [comment]
@justinshenk, would you like to address any/all of delve-team/delve#59 prior to concluding this review? A lot of the points there seem like fairly easy fixes...

**Comment by justinshenk** [comment]
Thanks guys for your patience! Corrections made as suggested in the latest version v0.1.49 as described in https://github.com/delve-team/delve/issues/59.
Latest Zenodo DOI for v0.1.49: https://zenodo.org/record/5865465.

**Comment by justinshenk** [comment]
@whedon set v0.1.49 as version

**Comment by whedon** [comment]
I'm sorry @justinshenk, I'm afraid I can't do that. That's something only editors are allowed to do.

**Comment by dionhaefner** [comment]
Alright, looking good! That's green light 🟢 from me. (CC @rkurchin)

**Comment by rkurchin** [comment]
@whedon set v0.1.49 as version

**Comment by whedon** [comment]
OK. v0.1.49 is the version.

**Comment by rkurchin** [comment]
@whedon set https://zenodo.org/record/5865465 as archive

**Comment by whedon** [comment]
https://zenodo.org/record/5865465 doesn't look like an archive DOI.

**Comment by rkurchin** [comment]
@whedon set 10.5281/zenodo.5865465 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5865465" target="_blank">10.5281/zenodo.5865465</a> is the archive.

**Comment by rkurchin** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by rkurchin** [comment]
A few editorial suggestions/remarks for the manuscript:
* line 44: should read "needs to be able to be used with as little computational and workflow overhead as possible" (first "as" currently missing)
* 48: "known" used twice, remove one of them
* 60: I would rephrase to "`Delve` directly hooks into..."
* 70-71: I would rephrase to "has exploded in recent years"
* 72: Pluralize "Publication"
* 76: "hinting on" should be "hinting at"
* 78: remove "an"
* 82-83: JFYI, it's not necessary to explicitly link the Zenodo here as it will be in the frontmatter of the paper. That being said, it's no problem if you want to!
* 104-105: Phrasing here is a bit confusing because you say you're going to compute Q(Z_l, Z_l) but then show an equation for Q(X,Y). I would rephrase this section to start with that equation and just say something like "We wish to compute the covariance between two random variables X and Y, given by..." and then remark that to avoid having to store all the samples in memory, we approximate by...etc.
* 105-106: I would add somewhere the formal definitions of x_i, y_i for completeness (I recognize it's probably obvious to most readers, but it still feels weird to me not to have it there)
* 119: "asynchronous" should be "asynchronously"
* 120: "numeric" should be "numerical"
* 121: the phrase "by default" should be moved to the end of the sentence (after "floating-point values")
* 126-127: I'm not sure what's happening notationally with x vs. dot, but I interpret both of those as multiplication, so some clarification is needed one why these are different.

**Comment by rkurchin** [comment]
@justinshenk reminder to finish this up, we're almost there! 🚀

**Comment by MLRichter** [comment]
https://github.com/delve-team/delve/pull/65

This PR should address all remaining issues.
Please briefly read through my changed made in line 104-105 regarding the covariance matrix and let me know if this satisfies you.

**Comment by MLRichter** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03992/joss.03992/10.21105.joss.03992.pdf) :page_facing_up: :point_left:

**Comment by rkurchin** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/wacv.2018.00097 is OK
- 10.1007/s11263-019-01228-7 is OK
- 10.5281/zenodo.5233860 is OK
- 10.1007/978-3-030-86340-1_11 is OK
- 10.1109/cvpr.2016.319 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by rkurchin** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/wacv.2018.00097 is OK
- 10.1007/s11263-019-01228-7 is OK
- 10.5281/zenodo.5233860 is OK
- 10.1007/978-3-030-86340-1_11 is OK
- 10.1109/cvpr.2016.319 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2905

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2905, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1487229082983092226) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2919
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03992](https://doi.org/10.21105/joss.03992)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
@sambitdash, @dionhaefner – many thanks for your reviews here and to @rkurchin for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@justinshenk – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03992/status.svg)](https://doi.org/10.21105/joss.03992)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03992">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03992/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03992/status.svg
   :target: https://doi.org/10.21105/joss.03992
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03992/status.svg)](https://doi.org/10.21105/joss.03992)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by justinshenk** [comment]
Thank you for all your guidance for this paper 🙏
